stablehlo/dialect/TypeInference.cpp

```diff
@@ -989,17 +990,31 @@ LogicalResult inferDynamicUpdateSliceOp(
   }
 
   // (C6)
-  for (auto [index, dims] : llvm::enumerate(
-           llvm::zip(operandType.getShape(), updateType.getShape()))) {
-    auto [operandDim, updateDim] = dims;
-    if (updateDim < 0 || updateDim > operandDim)
-      return emitOptionalError(location, "expects size at dimension ", index,
-                               " of update to be in range [0, ", operandDim,
-                               "]. Got: ", updateDim, ".");
-  }
+  if (operandType.hasRank() && updateType.hasRank())
+    for (auto [index, dims] : llvm::enumerate(
+             llvm::zip(operandType.getShape(), updateType.getShape()))) {
+      auto [operandDim, updateDim] = dims;
+      if (hlo::isDynamicDimSize(updateDim)) continue;
+      if (hlo::isStaticDimSize(operandDim)) {
+        if (updateDim < 0 || updateDim > operandDim)
+          return emitOptionalError(location, "expects size at dimension ",
+                                   index, " of update to be in range [0, ",
+                                   operandDim, "]. Got: ", updateDim, ".");
+      } else {
+        if (updateDim < 0)
+          return emitOptionalError(
+              location, "expects size at dimension ", index,
+              " of update to be non-negative. Got: ", updateDim, ".");
+      }
+    }
 
-  inferredReturnShapes.emplace_back(operandType.getShape(),
-                                    operandType.getElementType());
+  // (C1)
```

Comment:
Can we move (C1) to the top?

Reply:
It's tempting, but we probably shouldn't assume that the code upstream will clean up `inferredReturnShapes` if we end up failing verification. Let's keep C1 after all the checks, not the other way around.